fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexMaintainer.java

```diff
@@ -133,6 +133,20 @@ public Subspace getSecondarySubspace() {
     @Nonnull
     public abstract RecordCursor<IndexEntry> scanUniquenessViolations(@Nonnull TupleRange range, @Nullable byte[] continuation, @Nonnull ScanProperties scanProperties);
 
+    /**
+     * Validate entries in the index. It scans the index and checks if the record associated with each index entry exists.
+     * @param scanType the {@link IndexScanType type} of scan to perform
+     * @param range the range to validate
+     * @param continuation any continuation from a previous validation invocation
+     * @param scanProperties skip, limit and other properties of the validation
+     * @return a cursor over index entries that have no associated records.
+     */
+    @Nonnull
+    public abstract RecordCursor<IndexEntry> validateOrphanEntries(@Nonnull IndexScanType scanType,
+                                                                   @Nonnull TupleRange range,
+                                                                   @Nullable byte[] continuation,
+                                                                   @Nonnull ScanProperties scanProperties);
```

Comment:
I think we should leave the definition of validation up to the index maintainer itself.  That is, don't assume that the only validation that the maintainer can do is look for orphaned entries.  For example, it is very likely that a text index or a rank index could do additional validation and, in fact, some of those index entries may not even have the ability to point back to records in which case that are technically always orphaned (or determining if they are orphaned involves scanning the data rather than the rows).  

What I would suggest is that instead of returning `IndexEntry`'s, return something like `InvalidIndexEntry` that wraps the `IndexEntry` along with an enum or string that indicates that reason the entry is invalid.  Keep in mind that the validation could be happening in an index that doesn't live in the record layer, so you'd need to make sure that the reason is something extensible (I think can enum still works here though).

In addition, make sure that `InvalidIndexEntry` has a decent `toString()` so that maintainers can sub-class it to add additional methods and attributes, but each can turn itself into something that is suitable for logging.

Reply:
Also, technically this is a breaking change -- existing `IndexMaintainer`s, outside of this code base will fail to compiler without this method being implemented.   Personally, I think I like this because we really want new indexes to adopt this API.  @alecgrieser or @MMcM do you have strong opinions on this?

Reply:
It's probably okay if we increment the minor version and call it out in the release notes.

By rights, it should take two minor versions, one to change the status to anticipate the change and another to introduce it. But this may not be used widely enough to bother.

Reply:
I have one other breaking (ish) change I'd kind of like to make (#403). If we're bumping the minor version, I'd kind of like to get it in at the same time.

Reply:
@alecgrieser Sounds good. @dsimmen is also going to have one breaking change. Maybe we can take three at the same time.

If we are taking two minor versions, and if you are preparing the PR for the first bump, could you change `IndexMaintainer` to `UNSTABLE`?

Reply:
I think it's probably sufficient to only bump it once. It is technically in violation of our API stability contract, but it's probably not worth it, all things considered.

Reply:
Rebased to 2.6-dev